libcudacxx/include/cuda/std/detail/libcxx/include/__iterator/advance.h

```diff
@@ -28,45 +34,213 @@ _CCCL_IMPLICIT_SYSTEM_HEADER
 {
-    if (__n >= 0)
-        for (; __n > 0; --__n)
-            ++__i;
-    else
-        for (; __n < 0; ++__n)
-            --__i;
+  if (__n >= 0)
+  {
+    for (; __n > 0; --__n)
+    {
+      ++__i;
+    }
+  }
+  else
+  {
+    for (; __n < 0; ++__n)
+    {
+      --__i;
+    }
+  }
 }
 
 template <class _RandIter>
-inline _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_CONSTEXPR_AFTER_CXX14
-void __advance(_RandIter& __i,
-             typename iterator_traits<_RandIter>::difference_type __n, random_access_iterator_tag)
+_LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_CONSTEXPR_AFTER_CXX14 void
+__advance(_RandIter& __i, typename iterator_traits<_RandIter>::difference_type __n, random_access_iterator_tag)
 {
-   __i += __n;
+  __i += __n;
 }
 
-template <class _InputIter>
-inline _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_CONSTEXPR_AFTER_CXX14
-void advance(_InputIter& __i,
-             typename iterator_traits<_InputIter>::difference_type __n)
+template < class _InputIter,
+           class _Distance,
+           class _IntegralDistance = decltype(_CUDA_VSTD::__convert_to_integral(_CUDA_VSTD::declval<_Distance>())),
+           class                   = __enable_if_t<is_integral<_IntegralDistance>::value> >
+_LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_CONSTEXPR_AFTER_CXX14 void advance(_InputIter& __i, _Distance __orig_n)
 {
-    _LIBCUDACXX_ASSERT(__n >= 0 || __is_cpp17_bidirectional_iterator<_InputIter>::value,
-                       "Attempt to advance(it, -n) on a non-bidi iterator");
-    __advance(__i, __n, typename iterator_traits<_InputIter>::iterator_category());
+  typedef typename iterator_traits<_InputIter>::difference_type _Difference;
+  _Difference __n = static_cast<_Difference>(_CUDA_VSTD::__convert_to_integral(__orig_n));
+  _LIBCUDACXX_ASSERT(__n >= 0 || __is_cpp17_bidirectional_iterator<_InputIter>::value,
+                     "Attempt to advance(it, n) with negative n on a non-bidirectional iterator");
+  _CUDA_VSTD::__advance(__i, __n, typename iterator_traits<_InputIter>::iterator_category());
 }
 
 _LIBCUDACXX_END_NAMESPACE_STD
 
+#if _LIBCUDACXX_STD_VER > 14 && !defined(_LICBUDACXX_COMPILER_MSVC_2017)
+
+// [range.iter.op.advance]
+
+_LIBCUDACXX_BEGIN_NAMESPACE_RANGES
+_LIBCUDACXX_BEGIN_NAMESPACE_CPO(__advance)
+struct __fn
+{
+private:
+  template <class _Ip>
+  _LIBCUDACXX_INLINE_VISIBILITY static constexpr void __advance_forward(_Ip& __i, iter_difference_t<_Ip> __n)
+  {
+    while (__n > 0)
+    {
+      --__n;
+      ++__i;
+    }
+  }
+
+  template <class _Ip>
+  _LIBCUDACXX_INLINE_VISIBILITY static constexpr void __advance_backward(_Ip& __i, iter_difference_t<_Ip> __n)
+  {
+    while (__n < 0)
+    {
+      ++__n;
+      --__i;
+    }
+  }
+
+  template <class _Iter_difference>
+  _LIBCUDACXX_INLINE_VISIBILITY static constexpr auto
+  __magnitude_geq(_Iter_difference __a, _Iter_difference __b) noexcept
+  {
+    return __a == 0 ? __b == 0 : __a > 0 ? __a >= __b : __a <= __b;
```

Comment:
suggestion: a bit difficult to read, do you think something along the following lines could help?
```suggestion
    return __a == 0 ? __b == 0 : //
         __a > 0 ? __a >= __b
                 : __a <= __b;
```